### PR TITLE
feat: lookup manifest file size

### PR DESCRIPTION
**Pull request by QuenKar**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- use `HashMap<String, u64>` to keep track for every manifest file size in `ManifestObjectStore `.
- we can use `RegionManifestManager::manifest_size()` to get manifest files size.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/issues/2551

### CODE REVIEW

**Comment by QuenKar** [comment]
After discussing with @evenyag , we decide to only calculate manifest size from the latest checkpoint, the files'version <= latest checkpoint will not be included. 
There are two main reasons.
- manifest file size is usually not large
- there is no need to read the manifest file which version <= latest checkpoint

File: src/mito2/src/manifest/storage.rs [hunk 1]
```diff
@@ -134,6 +147,10 @@ pub struct ManifestObjectStore {
     object_store: ObjectStore,
     compress_type: CompressionType,
     path: String,
+    /// Stores the size of each manifest file.
+    /// K is the manifest file name(such as "0000000000000000006.checkpoint"
+    /// or "0000000000000000006.json"), and V is the file size.
+    manifest_size_map: HashMap<String, u64>,
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:153]: nitpick: Could we define an enum for the key? It should be more space-efficient than string.
```rust
/// Key to identify a manifest file.
enum FileKey {
    /// A delta file (`.json`).
    Delta(ManifestVersion),
    /// A checkpoint file (`.checkpoint`).
    Checkpoint(ManifestVersion),
}
```

File: src/mito2/src/manifest/storage.rs [hunk 2]
```diff
@@ -80,6 +80,19 @@ pub fn file_version(path: &str) -> ManifestVersion {
...
+/// ### Panics
+/// Panics if the file path is not a valid delta or checkpoint file.
+#[inline]
+pub fn file_name(path: &str) -> String {
+    let name = path.rsplit('/').next().unwrap_or("").to_string();
+    if !is_checkpoint_file(&name) && !is_delta_file(&name) {
+        panic!("Invalid file: {path}")
+    }
+    name
+}
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:94]: Prefer not to panic. You can return an `Option` and ignore that file if it isn't a valid file.

File: src/mito2/src/manifest/storage.rs [hunk 3]
```diff
@@ -424,6 +461,69 @@ impl ManifestObjectStore {
...
+    /// Get the size(Byte) of the checkpoint file by checkpoint version.
+    /// If the checkpoint file does not exist, return None.
+    pub fn checkpoint_file_size(&self, version: ManifestVersion) -> Option<u64> {
+        self.manifest_size_map
+            .get(&checkpoint_file(version))
+            .copied()
+    }
+
+    /// Compute the size(Byte) in manifest size map.
+    pub fn get_total_manifest_size(&self) -> u64 {
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:480]: style(nitpick): Remove `get_` and use the same method name as the manager.
```suggestion
    pub fn total_manifest_size(&self) -> u64 {
```

File: src/mito2/src/manifest/storage.rs [hunk 4]
```diff
@@ -80,6 +80,19 @@ pub fn file_version(path: &str) -> ManifestVersion {
     s.parse().unwrap_or_else(|_| panic!("Invalid file: {path}"))
 }
 
+/// Return's the file name from path
+/// Just use for .json and .checkpoint file
+/// ### Panics
+/// Panics if the file path is not a valid delta or checkpoint file.
+#[inline]
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:87]: No need to inline this function.
```suggestion
```

File: src/mito2/src/manifest/storage.rs [hunk 5]
```diff
@@ -296,13 +326,15 @@ impl ManifestObjectStore {
                 compress_type: self.compress_type,
                 path: &path,
             })?;
+        self.set_manifest_size_by_path(&path, data.len() as u64);
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:329]: We should set manifest size after `write()` returns `Ok`.
```rust
self.object_store.write().await?;
self.set_manifest_size_by_path(&path, data.len() as u64);
Ok(())
```

File: src/mito2/src/manifest/storage.rs [hunk 6]
```diff
@@ -424,6 +460,69 @@ impl ManifestObjectStore {
     pub async fn read_file(&self, path: &str) -> Result<Vec<u8>> {
         self.object_store.read(path).await.context(OpenDalSnafu)
     }
+
+    /// Get the size(Byte) of the delta file by delta version.
+    /// If the delta file does not exist, return None.
+    pub fn delta_file_size(&self, version: ManifestVersion) -> Option<u64> {
```
[the remark this answers is not in the document]
Comment (QuenKar) [reply src/mito2/src/manifest/storage.rs:491]: - `set_manifest_size_until()` has been deleted.
- `get_total_manifest_size()`,`set_manifest_size_by_path()`,`set_delta_file_size()` function are used in `manager.rs`. 
- `delta_file_size()`,`checkpoint_file_size()`,`set_checkpoint_file_size()` could be used in the future, but now they are not used yet. Maybe we can delete them.

[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/manifest/storage.rs:493]: Is `pub(crate)` enough?

File: src/mito2/src/manifest/storage.rs [hunk 7]
```diff
@@ -80,6 +80,16 @@ pub fn file_version(path: &str) -> ManifestVersion {
...
 
+/// Return the file name from path.
+/// Just for (`.json`) and (`.checkpoint`) file, other file will return None.
+pub fn file_name(path: &str) -> Option<String> {
+    let name = path.rsplit('/').next().unwrap_or("").to_string();
+    if !is_checkpoint_file(&name) && !is_delta_file(&name) {
+        return None;
+    }
+    Some(name)
+}
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:91]: Looks like the function is unused.

File: src/mito2/src/manifest/storage.rs [hunk 8]
```diff
@@ -264,12 +291,15 @@ impl ManifestObjectStore {
...
             .collect();
+        let paths = del_entries
+            .iter()
+            .map(|(e, _, _)| e.path().to_string())
+            .collect::<Vec<_>>();
         let ret = paths.len();
 
         debug!(
-            "Deleting {} logs from manifest storage path {} until {}, checkpoint: {:?}, paths: {:?}",
+            "Deleting {} logs from manifest storage path {} until {}, checkpoint_version: {:?}, paths: {:?}",
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:302]:
```suggestion
            "Deleting {} logs from manifest storage path {} until {}, checkpoint_version: {}, paths: {:?}",
```

File: src/mito2/src/manifest/storage.rs [hunk 9]
```diff
@@ -282,10 +312,21 @@ impl ManifestObjectStore {
             .await
             .context(OpenDalSnafu)?;
 
+        // delete the manifest'size
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:315]:
```suggestion
        // delete manifest sizes
```

File: src/mito2/src/manifest/storage.rs [hunk 10]
```diff
@@ -351,12 +398,15 @@ impl ManifestObjectStore {
         let checkpoint_data =
             match self.object_store.read(&path).await {
                 Ok(checkpoint) => {
+                    let checkpoint_size = checkpoint.len();
                     let decompress_data = self.compress_type.decode(checkpoint).await.context(
                         DecompressObjectSnafu {
                             compress_type: self.compress_type,
-                            path,
+                            path: path.clone(),
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:405]: Do we still need to clone the path?

File: src/mito2/src/manifest/storage.rs [hunk 11]
```diff
@@ -373,13 +423,15 @@ impl ManifestObjectStore {
...
                             match self.object_store.read(&fall_back_path).await {
                                 Ok(checkpoint) => {
+                                    let checkpoint_size = checkpoint.len();
                                     let decompress_data = FALL_BACK_COMPRESS_TYPE
                                         .decode(checkpoint)
                                         .await
                                         .context(DecompressObjectSnafu {
                                             compress_type: FALL_BACK_COMPRESS_TYPE,
-                                            path,
+                                            path: path.clone(),
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:432]: Do we still need to clone the path?

File: src/mito2/src/manifest/storage.rs (hunk 8 above)
Comment (QuenKar) [reply src/mito2/src/manifest/storage.rs:302]: `checkpoint_version` is `Option<&u64>` which doesn't impl `Display` trait.

**Review by evenyag — APPROVED** [review]

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**